File: algocoding/leetcode/two_sum.cpp
/*******************************************************************************
                             Two Sum(Medium)
 *******************************************************************************
 Given an array of integers, find two numbers such that they add up to a specific 
 target number.
 
 The function twoSum should return indices of the two numbers such that they add 
 up to the target, where index1 must be less than index2. Please note that your 
 returned answers (both index1 and index2) are not zero-based.
 
 You may assume that each input would have exactly one solution.
 
 Input: numbers={2, 7, 11, 15}, target=9
 Output: index1=1, index2=2
 *******************************************************************************
                             interface
 *******************************************************************************
 class Solution {
 public:
	 vector<int> twoSum(vector<int> &numbers, int target) {
		 
	 }
 };
 *******************************************************************************
                              algo
 *******************************************************************************
  
 ******************************************************************************/
#include <iostream>
#include <string>
//#include <unordered_map> /* g++ -std=c++11 */
#include <map>
#include <vector>
using namespace std;

class Solution {


public:
vector < int > twoSum(vector < int > &numbers, int target) 
{
    map<int, int> value_map; /*value:index*/
    vector<int> result;
	int i,gap;
	
    for (i = 0; i < numbers.size(); i++)
        value_map[numbers[i]] = i;

    for (i = 0; i < numbers.size(); i++) 
	{
        gap = target - numbers[i];
		
        if (value_map.find(gap) != value_map.end() && value_map[gap] > i) 
		{
            result.push_back(i+1);/* index from 1 */
            result.push_back(value_map[gap]+1);
            break;
        }
    }
	
    return result;		 
}




};


void two_sum_test01()
{
    vector<int> v;
    vector<int> ret;
	int target;
	Solution solu;

    v.push_back(49);
    v.push_back(50);
    v.push_back(51);
    v.push_back(99);
    v.push_back(101);
	
	target = 100;

	ret = solu.twoSum(v,target);

    for (unsigned int i = 0; i < ret.size( ); i++)
        cout << "index: " << ret[i]-1 << " value: " << v[ret[i]-1] << endl;	
}

int main()
{
    two_sum_test01();
}

File: algocoding/leetcode/java/twoSum.java
// http://www.cnblogs.com/springfor/p/3859618.html

public int[] twoSum(int[] numbers, int target) {
        int [] res = new int[2];
        if(numbers==null||numbers.length<2)
            return res;
        
        //copy original list and sort
        int[] copylist = new int[numbers.length];  
        System.arraycopy(numbers, 0, copylist, 0, numbers.length);  
        Arrays.sort(copylist);    
        
        int low = 0;
        int high = copylist.length-1;
        
        while(low<high){
            if(copylist[low]+copylist[high]<target)
                low++;
            else if(copylist[low]+copylist[high]>target)
                high--;
            else{
                res[0]=copylist[low];
                res[1]=copylist[high];
                break;
            }
        }
        
        //find index from original list
        int index1 = -1, index2 = -1;  
        for(int i = 0; i < numbers.length; i++){  
            if(numbers[i] == res[0]&&index1==-1)
                index1 = i+1;
            else if(numbers[i] == res[1]&&index2==-1)
                index2 = i+1;
       } 
        res[0] = index1;
        res[1] = index2;
        Arrays.sort(res);
        return res;
    }

